logbook-json/src/main/java/org/zalando/logbook/json/JsonHttpLogFormatter.java

```diff
@@ -1,92 +1,169 @@
 package org.zalando.logbook.json;
 
-import com.fasterxml.jackson.annotation.JsonRawValue;
-import com.fasterxml.jackson.annotation.JsonValue;
-import com.fasterxml.jackson.databind.ObjectMapper;
-import lombok.AllArgsConstructor;
-import org.apiguardian.api.API;
-import org.zalando.logbook.HttpLogFormatter;
-import org.zalando.logbook.HttpMessage;
-import org.zalando.logbook.StructuredHttpLogFormatter;
+import static org.apiguardian.api.API.Status.STABLE;
 
 import java.io.IOException;
+import java.io.StringWriter;
+import java.util.List;
 import java.util.Map;
 import java.util.Optional;
 
-import static org.apiguardian.api.API.Status.STABLE;
+import org.apiguardian.api.API;
+import org.zalando.logbook.Correlation;
+import org.zalando.logbook.HttpLogFormatter;
+import org.zalando.logbook.HttpMessage;
+import org.zalando.logbook.HttpRequest;
+import org.zalando.logbook.HttpResponse;
+import org.zalando.logbook.Origin;
+import org.zalando.logbook.Precorrelation;
+
+import com.fasterxml.jackson.core.JsonFactory;
+import com.fasterxml.jackson.core.JsonGenerator;
+import com.fasterxml.jackson.databind.ObjectMapper;
 
 /**
- * A custom {@link HttpLogFormatter} that produces JSON objects. It can be augmented with composition:
- *
- * <pre>
- * {@code
- *
- * public class CustomsFormatter implements HttpLogFormatter {
- *
- *     private final JsonHttpLogFormatter delegate;
- *
- *     public CustomsFormatter(final ObjectMapper mapper) {
- *         this.delegate = new JsonHttpLogFormatter(mapper);
- *     }
- *
- *     public String format(final Precorrelation<HttpRequest> precorrelation) throws IOException {
- *         Map<String, Object> request = delegate.prepare(precorrelation);
- *         // modify request here
- *         return delegate.format(request);
- *     }
- *
- *     public String format(final Correlation<HttpRequest, HttpResponse> correlation) throws IOException {
- *         Map<String, Object> response = delegate.prepare(correlation);
- *         // modify response here
- *         return delegate.format(response);
- *      }
- *
- * }
- * }
- * </pre>
+ * A custom {@link HttpLogFormatter} that produces JSON objects. 
  */
 @API(status = STABLE)
-public final class JsonHttpLogFormatter implements StructuredHttpLogFormatter {
```

Comment:
Technically that's a breaking change since the public API of the class changed.

Reply:
You're technically correct, however

 * StructuredHttpLogFormatter is still status experimental, so JsonHttpLogFormatter should be too?
 * You're about to do a new major version, some API changes are to be expected?
 * The class is final, so there exist no subclasses which would (potentially silently) break?

So can I leave it as is?

Reply:
> * StructuredHttpLogFormatter is still status experimental, so JsonHttpLogFormatter should be too?
> * You're about to do a new major version, some API changes are to be expected?

Both true.

> * The class is final, so there exist no subclasses which would (potentially silently) break?

Those methods are public and the preferred usage is composition where users are allowed to call the public methods, e.g. by augmenting the map before passing it down.

Reply:
So your point is that those who now change (add / remove / transform) the output JSON format are now unable to do so using the same JsonHttpLogFormatter. 

Letting JsonHttpLogFormatter implement StructuredHttpLogFormatter is still possible, then HttpLogFormatter#format methods run the optimized code, while the StructuredHttpLogFormatter#prepare / #format still runs as today. The will be some small overlap in functionality, but not that much really.

Reply:
> The will be some small overlap in functionality, but not that much really.

Overlap in functionality without overlap in code, i.e. there is no reuse between the two approaches. In that case I'd actually rather split those two use cases up into two classes (banking on the fact that we're going to release a major version anyway, so breaking changes are easy right now).